### PR TITLE
New-style parameter changes and options tab update

### PR DESCRIPTION
**Pull request by Jannes123**
@timlinux and @lucernae please review.  There is one test build_widget which I was unable to update @lucernae maybe you can take a look at it it uses your input_list_parameter for  testing.  There is alot of maintenance needed still but I think it is a worth while commit.

### CODE REVIEW

File: safe/impact_functions/earthquake/itb_earthquake_fatality_model/parameter_definitions.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+"""
+
+from safe_extras.parameters.boolean_parameter import BooleanParameter
+from safe_extras.parameters.string_parameter import StringParameter
+from safe_extras.parameters.list_parameter import ListParameter
+from safe_extras.parameters.float_parameter import FloatParameter
+from safe_extras.parameters.dict_parameter import DictParameter
+
+
+def x_coefficient():
```
Comment (timlinux) [inline safe/impact_functions/earthquake/itb_earthquake_fatality_model/parameter_definitions.py]: I thought we agreed that eq params would be hard coded into the IF and not made available to the end user? (applied to all below).

File: safe/impact_functions/earthquake/pager_earthquake_fatality_model/impact_function.py [hunk 2]
```diff
@@ -69,8 +53,8 @@ def fatality_rate(self, mmi):
         """
 
         N = math.sqrt(2 * math.pi)
-        THETA = self.hardcoded_parameters['Theta']
-        BETA = self.hardcoded_parameters['Beta']
+        THETA = self.parameters['Theta'].value
```
Comment (timlinux) [inline safe/impact_functions/earthquake/pager_earthquake_fatality_model/impact_function.py]: Again these should stay hard coded.

File: safe/impact_functions/earthquake/pager_earthquake_fatality_model/parameter_definitions.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+from safe_extras.parameters.boolean_parameter import BooleanParameter
+from safe_extras.parameters.dict_parameter import DictParameter
+from safe_extras.parameters.float_parameter import FloatParameter
+from safe_extras.parameters.input_list_parameter import InputListParameter
+
+__author__ = 'lucernae'
+__date__ = '13/04/15'
+
+
+def theta():
```
Comment (timlinux) [inline safe/impact_functions/earthquake/pager_earthquake_fatality_model/parameter_definitions.py]: As above - these must be hard coded and not user editable

File: safe/impact_functions/generic/continuous_hazard_population/metadata_definitions.py [hunk 4]
```diff
@@ -100,7 +102,7 @@ def as_dict():
             },
             # Configurable parameters
             'parameters': OrderedDict([
-                ('Categorical thresholds', [0.34, 0.67, 1]),
+                ('Categorical thresholds', categorical_thresholds()),
```
Comment (timlinux) [inline safe/impact_functions/generic/continuous_hazard_population/metadata_definitions.py]: This should be explicitly named (i.e. should have flood in the name). I am also confused as to why this parameter is not called ``continuous_flood_threshold`` and its title ``tr('Flood threshohlds'))``

File: safe/impact_functions/generic/continuous_hazard_population/parameter_definitions.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# coding=utf-8
+from safe_extras.parameters.input_list_parameter import InputListParameter
+
+__author__ = 'lucernae'
+__date__ = '13/04/15'
+
+
+def categorical_thresholds():
```
Comment (timlinux) [inline safe/impact_functions/generic/continuous_hazard_population/parameter_definitions.py]: Again the semantics here seem wrong - we are working with continuous data so using the overloaded term 'Categorical' here is not good.

File: safe/impact_functions/inundation/flood_polygon_population/parameter_definitions.py [hunk 6]
```diff
@@ -13,6 +13,22 @@
 
 from safe_extras.parameters.boolean_parameter import BooleanParameter
 from safe_extras.parameters.string_parameter import StringParameter
+from safe_extras.parameters.float_parameter import FloatParameter
+
+
+def evacuation_percentage():
+    """Generator for the default evaluation percentage parameter.
```
Comment (timlinux) [inline safe/impact_functions/inundation/flood_polygon_population/parameter_definitions.py]: Please leave newline between docstring header and params

File: safe/impact_functions/inundation/flood_polygon_population/parameter_definitions.py [hunk 7]
```diff
@@ -45,7 +65,11 @@ def affected_field():
 
 
 def affected_value():
-    """Generator for parameter stating what values constitute 'affected'."""
+    """Generate affected value parameter
```
Comment (timlinux) [inline safe/impact_functions/inundation/flood_polygon_population/parameter_definitions.py]: Please ensure docstring first line is a complete sentance (terminated with a full stop). Applies to all docstrings above too.

File: safe/impact_functions/inundation/flood_polygon_population/parameter_definitions.py [hunk 8]
```diff
@@ -45,7 +65,11 @@ def affected_field():
...
+
+    :return: list of String Parameter
+    :rtype: list[StringParameter]
```
Comment (timlinux) [inline safe/impact_functions/inundation/flood_polygon_population/parameter_definitions.py]: I think this should just be list - please check in pycharm if it correctly interprets the rtype if you specify a type in the list.

File: safe/impact_functions/inundation/flood_polygon_population/test/test_flood_polygon_population.py [hunk 9]
```diff
@@ -49,8 +50,19 @@ def test_run(self):
 
         function.hazard = hazard_layer
         function.exposure = exposure_layer
-        function.parameters['affected_field'] = 'FLOODPRONE'
-        function.parameters['affected_value'] = 'YES'
+
+        field1 = StringParameter()
```
Comment (timlinux) [inline safe/impact_functions/inundation/flood_polygon_population/test/test_flood_polygon_population.py]: As discussed in our meeting, this should be expressed in KW metadata for the flood layer rather.

File: safe/impact_functions/inundation/flood_polygon_population/test/test_flood_polygon_population.py [hunk 10]
```diff
@@ -49,8 +50,19 @@ def test_run(self):
...
+        field1.name = 'Affected Field'
+        field1.is_required = True
+        field1.value = 'FLOODPRONE'
+        function.parameters['affected_field'] = field1
+
+        field2 = StringParameter()
```
Comment (timlinux) [inline safe/impact_functions/inundation/flood_polygon_population/test/test_flood_polygon_population.py]: Also should be in keywords rather.

File: safe/impact_functions/inundation/flood_polygon_roads/impact_function.py [hunk 11]
```diff
@@ -71,12 +71,12 @@ def run(self, layers=None):
...
 
         # Set the target field
-        target_field = 'FLOODED'
+        target_field = self.parameters['target_field_value'].value
 
         # Get the parameters from IF options
-        road_type_field = self.parameters['road_type_field']
-        affected_field = self.parameters['affected_field']
-        affected_value = self.parameters['affected_value']
+        road_type_field = self.parameters['road_type_field'].value
```
Comment (timlinux) [inline safe/impact_functions/inundation/flood_polygon_roads/impact_function.py]: As above we want to move this logic to more like this:

```
self.hazard_metadata.type_field
self.exposure_metadata.affected_field
self.exposure_metadata.affected_value
```

Maybe we must merge this as you have it for now and do the above in a second pass, but please liaise with @akbargumbira and @ismailsunni to sequence this work.

File: safe/impact_functions/inundation/flood_polygon_roads/metadata_definitions.py [hunk 12]
```diff
@@ -79,13 +85,14 @@ def as_dict():
...
                 # information about road types
-                ('road_type_field', 'TYPE'),
+                ('target_field_value', target_field()),
+                ('road_type_field', road_type_field()),
                 # This field of the  hazard layer contains information
                 # about inundated areas
-                ('affected_field', 'affected'),
+                ('affected_field', affected_field()),
                 # This value in 'affected_field' of the hazard layer
                 # marks the areas as inundated
```
Comment (timlinux) [inline safe/impact_functions/inundation/flood_polygon_roads/metadata_definitions.py]: Identical comments above about moving to keywords for these.

File: safe/impact_functions/inundation/flood_raster_osm_building_impact/parameter_definitions.py [hunk 13]
```diff
@@ -66,3 +67,30 @@ def building_type_field():
     field.is_required = True
     field.value = True
     return field
+
+
+def threshold():
```
Comment (timlinux) [inline safe/impact_functions/inundation/flood_raster_osm_building_impact/parameter_definitions.py]: ``thresholds`` since we return a list

File: safe/impact_functions/inundation/flood_raster_osm_building_impact/parameter_definitions.py [hunk 14]
```diff
@@ -66,3 +67,30 @@ def building_type_field():
...
+    """
+    field = FloatParameter()
+    field.name = 'Thresholds [m]'
+    field.is_required = True
+    field.precision = 2
+    field.value = 1.0  # default value
+    return field
+
+
+def hazard_level_name():
```
Comment (timlinux) [inline safe/impact_functions/inundation/flood_raster_osm_building_impact/parameter_definitions.py]: plural as it returns a list -> hazard_level_names

File: safe/impact_functions/inundation/flood_raster_road_qgis/parameter_definitions.py [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+__author__ = 'lucernae'
+__date__ = '11/04/15'
+
+
+import sys
+from safe_extras.parameters.float_parameter import FloatParameter
+from safe_extras.parameters.string_parameter import StringParameter
+
+
+def road_type_field():
```
Comment (timlinux) [inline safe/impact_functions/inundation/flood_raster_road_qgis/parameter_definitions.py]: to keywords

File: safe/impact_functions/inundation/flood_raster_road_qgis_gdal/parameter_definitions.py [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+__author__ = 'lucernae'
+__date__ = '11/04/15'
+
+
+import sys
+from safe_extras.parameters.float_parameter import FloatParameter
+from safe_extras.parameters.string_parameter import StringParameter
+
+
+def road_type_field():
```
Comment (timlinux) [inline safe/impact_functions/inundation/flood_raster_road_qgis_gdal/parameter_definitions.py]: to keywords

File: safe/impact_functions/inundation/flood_vector_building_impact/impact_function.py [hunk 17]
```diff
@@ -51,8 +51,8 @@ def notes(self):
         :return: The notes that should be attached to this impact report.
         :rtype: list
         """
-        affected_field = self.parameters['affected_field']
-        affected_value = self.parameters['affected_value']
+        affected_field = self.parameters['affected_field'].value
```
Comment (timlinux) [inline safe/impact_functions/inundation/flood_vector_building_impact/impact_function.py]: to keywords

**Comment by timlinux** [comment]
Hi Jannes

I made some inline comments but I can summarise my comments in 3 categories:

## Coding standards

* Docstrings should following the standards defined here: https://github.com/AIFDR/inasafe/wiki/Coding-Standards#doc-strings-and-comments
* Parameter names should be pluralised when they represent a list

## Hard coded parameters

* In particular for the EQ functions, the parameters should not be exposed to end users, they should rather be hard coded as IF constants

## Layer attributes

* As per our scrum meeting, layer attributes should not be stored in IF, but rather retrieved by IF from layer keywords. However in the interests of being pragmatic, lets include them now in this changeset and then move to remove them from all IF's in a future PR.

In summary, please address the first two items above, then LGTM

**Comment by lucernae** [comment]
As for me, I haven't addressed the details of some of the parameter (like the description or help text or precision settings), because it was vague for me at that time, and I only focused on making these connects. 
So, I assume it will be good if we have detailed description in the IF options, just like what we have in the needs profile.
Also, we need to enclose description or title parameters inside tr() so we can translate it.

**Comment by Jannes123** [comment]
Thanks @lucernae  I was just about to ask for help on this.  If we cant merge the whole thing let me know maybe we can break it up into smaller commits.

**Comment by lucernae** [comment]
@Jannes123 I apologize I didn't ask your permissions earlier. There are some changes on the IF options (what to and what not to include in the IF options) and I think I can help since I can discuss it face to face with @akbargumbira, @ismailsunni and @timlinux at AIFDR office :)

**Comment by Jannes123** [comment]
No problem I'm just glad the parameters rework is not going stale.
